main/src/ca/uwaterloo/flix/language/phase/Verifier.scala

```diff
@@ -40,42 +42,44 @@ object Verifier {
-      case Constant.Float32(_) => check(expected = MonoType.Float32)(actual = tpe, loc)
-      case Constant.Float64(_) => check(expected = MonoType.Float64)(actual = tpe, loc)
-      case Constant.BigDecimal(_) => check(expected = MonoType.BigDecimal)(actual = tpe, loc)
-      case Constant.Int8(_) => check(expected = MonoType.Int8)(actual = tpe, loc)
-      case Constant.Int16(_) => check(expected = MonoType.Int16)(actual = tpe, loc)
-      case Constant.Int32(_) => check(expected = MonoType.Int32)(actual = tpe, loc)
-      case Constant.Int64(_) => check(expected = MonoType.Int64)(actual = tpe, loc)
-      case Constant.BigInt(_) => check(expected = MonoType.BigInt)(actual = tpe, loc)
-      case Constant.Str(_) => check(expected = MonoType.String)(actual = tpe, loc)
-      case Constant.Regex(_) => check(expected = MonoType.Regex)(actual = tpe, loc)
-    }
+  private def visitExpr(expr: Expr)(implicit root: Root, env: Map[Symbol.VarSym, MonoType], lenv: Map[Symbol.LabelSym, MonoType]): (MonoType, Purity) = expr match {
+
+    case Expr.Cst(cst, tpe, loc) =>
+      val t = cst match {
+        case Constant.Unit => check(expected = MonoType.Unit)(actual = tpe, loc)
+        case Constant.Null => tpe
+        case Constant.Bool(_) => check(expected = MonoType.Bool)(actual = tpe, loc)
+        case Constant.Char(_) => check(expected = MonoType.Char)(actual = tpe, loc)
+        case Constant.Float32(_) => check(expected = MonoType.Float32)(actual = tpe, loc)
+        case Constant.Float64(_) => check(expected = MonoType.Float64)(actual = tpe, loc)
+        case Constant.BigDecimal(_) => check(expected = MonoType.BigDecimal)(actual = tpe, loc)
+        case Constant.Int8(_) => check(expected = MonoType.Int8)(actual = tpe, loc)
+        case Constant.Int16(_) => check(expected = MonoType.Int16)(actual = tpe, loc)
+        case Constant.Int32(_) => check(expected = MonoType.Int32)(actual = tpe, loc)
+        case Constant.Int64(_) => check(expected = MonoType.Int64)(actual = tpe, loc)
+        case Constant.BigInt(_) => check(expected = MonoType.BigInt)(actual = tpe, loc)
+        case Constant.Str(_) => check(expected = MonoType.String)(actual = tpe, loc)
+        case Constant.Regex(_) => check(expected = MonoType.Regex)(actual = tpe, loc)
+      }
+      (t, Purity.Pure)
 
     case Expr.Var(sym, tpe1, loc) => env.get(sym) match {
       case None => throw InternalCompilerException(s"Unknown variable sym: '$sym'", sym.loc)
       case Some(tpe2) =>
-        checkEq(tpe1, tpe2, loc)
+        (checkEq(tpe1, tpe2, loc), Purity.Pure)
     }
 
-    case Expr.ApplyAtomic(op, exps, tpe, _, loc) =>
-      val ts = exps.map(visitExpr)
+    case Expr.ApplyAtomic(op, exps, tpe, pur, loc) =>
+      val rs = exps.map(visitExpr)
 
       op match {
         case AtomicOp.Unary(sop) =>
-          val List(t) = ts
+          val (List(t), List(p)) = rs.unzip
           val opTpe = sop match {
             case SemanticOp.BoolOp.Not => MonoType.Bool
```

Comment:
In general, if it doesn't hurt the code to much, it would be nice that in a given expression it is clear what its inherent effect is and what the "nested" effect is